book/src/plumbing.md

```diff
@@ -0,0 +1,279 @@
+# Plumbing
+
+This chapter documents the code that salsa generates and its "inner workings".
+We refer to this as the "plumbing".
+
+This page walks through the ["Hello, World!"] example and explains the code that
+it generates.
+
+["Hello, World!"]: https://github.com/salsa-rs/salsa/blob/master/examples/hello_world/main.rs
+
+If you'd like to see for yourself, you can set the environment variable
+`SALSA_DUMP` to 1 while the procedural macro runs, and it will dump the full
+output to stdout. I recommend piping the output through rustfmt.
+
+## Query groups and query group structs
+
+When you define a query group trait:
+
+```rust,no_run,no_playground
+{{#include ../../examples/hello_world/main.rs:trait}}
+```
+
+the `salsa::query_group` macro generates a number of things:
+
+* a copy of the `HelloWorld` trait, minus the salsa annotations, and lightly edited
+* a "group struct" named `HelloWorldStorage` that represents the group; this struct implements `plumbing::QueryGroup`
+  * somewhat confusingly, this struct doesn't actually contain the storage itself, but rather has an associated type that leads to the "true" storage struct
+* an impl of the `HelloWorld` trait, for any database type
+* for each query, a "query struct" named after the query; these structs implement `plumbing::Query` and sometimes other plumbing traits
+* a group key, an enum that can identify any query within the group and store its key
+* the associated storage struct, which contains the actual hashmaps that store the data for all queries in the group
+
+Note that there are a number of structs and types (e.g., the query structs, or
+the group descriptor) that represent things which don't have "public" names. We
+currently generate mangled names but those names are not meant to be exposed to
```

Comment:
Not exactly true, QueryStructs are a part of public API. We use them, eg, to measure memory: https://github.com/rust-analyzer/rust-analyzer/blob/dff62def2ea31e7eb0b010c35d60cd53318eab80/crates/ra_ide_db/src/change.rs#L167-L269

Reply:
Well, I think it's true that this is in violation of our semver rules. It may also be true that we should expose some way to do this that we *do* want to commit to.

Reply:
(Or just commit to the current way.)

Reply:
I rewrote the text to be a bit ambiguous here. I'd be ok with "committing" to these structs existing, at least, and maybe (I can't remember...) we already did, though in that case we should probably document it more clearly.

Reply:
Well, I guess we basically did commit to that, what with the `db.query(Q)` methods.